### PR TITLE
Automatically give access to catalogs and releases to users with org access

### PR DESCRIPTION
**Pull request by anvddriesch**
Towards https://github.com/giantswarm/giantswarm/issues/20914

This adds a resource that binds the roles for releases and default catalog access dynamically.

## Checklist

- [ ] Update changelog in CHANGELOG.md.
- [ ] If the change introduces new roles, please consider adding user-friendly descriptions to the roles with the annotation `giantswarm.io/notes` to help explain their purpose and usage.

### CODE REVIEW

File: service/controller/rbac/resource/externalresources/create.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	}
+	orgNamespace := ns.Name
+
+	// List roleBindings in org-namespace
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("Listing RoleBindings in namespace %s.", orgNamespace))
+	orgRoleBindings, err := r.k8sClient.RbacV1().RoleBindings(orgNamespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	} else if len(orgRoleBindings.Items) == 0 {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("No RoleBindings found in namespace %s.", orgNamespace))
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/create.go:36]: Likewise, I would remove this log line

File: service/controller/rbac/resource/externalresources/create.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Resource) createOrUpdateClusterRoleBinding(ctx context.Context, clusterrolebinding *rbacv1.ClusterRoleBinding) error {
+	existingClusterRoleBinding, err := r.k8sClient.RbacV1().ClusterRoleBindings().Get(ctx, clusterrolebinding.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating clusterrolebinding %#q", clusterrolebinding.Name))
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/create.go:123]:
```suggestion
		r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("creating clusterrolebinding %#q", clusterrolebinding.Name))
```

File: service/controller/rbac/resource/externalresources/create.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+		_, err := r.k8sClient.RbacV1().ClusterRoleBindings().Create(ctx, clusterrolebinding, metav1.CreateOptions{})
+		if apierrors.IsAlreadyExists(err) {
+			// do nothing
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("clusterrolebinding %#q has been created", clusterrolebinding.Name))
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/create.go:132]:
```suggestion
		r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("clusterrolebinding %#q has been created", clusterrolebinding.Name))
```

File: service/controller/rbac/resource/externalresources/create.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+	} else if err != nil {
+		return microerror.Mask(err)
+	} else if clusterRoleBindingNeedsUpdate(clusterrolebinding, existingClusterRoleBinding) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updating clusterrolebinding %#q", clusterrolebinding.Name))
+		_, err := r.k8sClient.RbacV1().ClusterRoleBindings().Update(ctx, clusterrolebinding, metav1.UpdateOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("clusterrolebinding %#q has been updated", clusterrolebinding.Name))
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/create.go:142]:
```suggestion
		r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("clusterrolebinding %#q has been updated", clusterrolebinding.Name))
```

File: service/controller/rbac/resource/externalresources/create.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	} else {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("clusterrolebinding %#q already exists.", clusterrolebinding.Name))
+	}
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/create.go:145]:
```suggestion
	}
```

File: service/controller/rbac/resource/externalresources/create.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+	return nil
+}
+
+func (r *Resource) createOrUpdateRoleBinding(ctx context.Context, namespace string, roleBinding *rbacv1.RoleBinding) error {
+	existingRoleBinding, err := r.k8sClient.RbacV1().RoleBindings(namespace).Get(ctx, roleBinding.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating rolebinding %#q.", roleBinding.Name))
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/create.go:152]:
```suggestion
		r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("creating rolebinding %#q in namespace %s", roleBinding.Name, namespace))
```

File: service/controller/rbac/resource/externalresources/create.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+		_, err := r.k8sClient.RbacV1().RoleBindings(namespace).Create(ctx, roleBinding, metav1.CreateOptions{})
+		if apierrors.IsAlreadyExists(err) {
+			// do nothing
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("rolebinding %#q has been created.", roleBinding.Name))
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/create.go:161]:
```suggestion
		r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("rolebinding %#q in namespace %s has been created.", roleBinding.Name, namespace))
```

File: service/controller/rbac/resource/externalresources/create.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+	} else if err != nil {
+		return microerror.Mask(err)
+	} else if roleBindingNeedsUpdate(roleBinding, existingRoleBinding) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updating rolebinding %#q.", roleBinding.Name))
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/create.go:166]:
```suggestion
		r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("updating rolebinding %#q in namespace %s", roleBinding.Name, namespace))
```

File: service/controller/rbac/resource/externalresources/create.go [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+		_, err := r.k8sClient.RbacV1().RoleBindings(namespace).Update(ctx, roleBinding, metav1.UpdateOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("rolebinding %#q has been updated.", roleBinding.Name))
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/create.go:171]:
```suggestion
		r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("rolebinding %#q in namespace %s has been updated.", roleBinding.Name, namespace))
```

File: service/controller/rbac/resource/externalresources/delete.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+func (r *Resource) deleteClusterRoleBinding(ctx context.Context, clusterRoleBinding string) error {
+	var err error
+
+	_, err = r.k8sClient.RbacV1().ClusterRoleBindings().Get(ctx, clusterRoleBinding, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// nothing to be done
+	} else if err != nil {
+		return microerror.Mask(err)
+	} else {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("Deleting %#q clusterRoleBinding.", clusterRoleBinding))
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/delete.go:47]:
```suggestion
		r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("deleting  clusterrolebinding %#q", clusterRoleBinding))
```

File: service/controller/rbac/resource/externalresources/delete.go [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+		err = r.k8sClient.RbacV1().ClusterRoleBindings().Delete(ctx, clusterRoleBinding, metav1.DeleteOptions{})
+		if apierrors.IsNotFound(err) {
+			// nothing to be done
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("ClusterRoleBinding %#q has been deleted.", clusterRoleBinding))
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/delete.go:55]:
```suggestion
		r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("clusterrolebinding %#q has been deleted", clusterRoleBinding))
```

File: service/controller/rbac/resource/externalresources/delete.go [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+func (r *Resource) deleteRoleBinding(ctx context.Context, namespace string, roleBinding string) error {
+	var err error
+
+	_, err = r.k8sClient.RbacV1().RoleBindings(namespace).Get(ctx, roleBinding, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// nothing to be done
+	} else if err != nil {
+		return microerror.Mask(err)
+	} else {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("Deleting %#q roleBinding.", roleBinding))
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/delete.go:69]:
```suggestion
		r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("deleting rolebinding %#q from namespace %s", roleBinding, namespace))
```

File: service/controller/rbac/resource/externalresources/delete.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+		err = r.k8sClient.RbacV1().RoleBindings(namespace).Delete(ctx, roleBinding, metav1.DeleteOptions{})
+		if apierrors.IsNotFound(err) {
+			// nothing to be done
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("RoleBinding %#q has been deleted.", roleBinding))
```
Comment (marians) [inline service/controller/rbac/resource/externalresources/delete.go:77]:
```suggestion
		r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("rolebinding %#q has been deleted from namespace %s", roleBinding, namespace))
```

**Comment by marians** [comment]
I'm done here

**Comment by anvddriesch** [comment]
:smiley: I wanted to change the logs to the new format before requesting review but I guess it just happened automatically. But thank you for already taking care of it.

**Comment by anvddriesch** [comment]
I tested this on `gauss`.
- created a new org
- gave read-all and read-cluster-in-ns permissions to new test user in the org
- the rolebindings giving read access for catalogs and releases were created
- using impersonation I could see those resources as well
- on deletion of the org, the created bindings were deleted.

**Review by marians — APPROVED** [review]

**Pull request merged**